packages/core/contracts/financial-templates/optimistic-distributor/OptimisticDistributor.sol

```diff
@@ -0,0 +1,420 @@
+                rewardToken: rewardToken,
+                maximumRewardAmount: maximumRewardAmount,
+                earliestProposalTimestamp: earliestProposalTimestamp,
+                optimisticOracleProposerBond: optimisticOracleProposerBond,
+                optimisticOracleLivenessTime: optimisticOracleLivenessTime,
+                priceIdentifier: priceIdentifier,
+                customAncillaryData: customAncillaryData
+            });
+        uint256 rewardIndex = nextCreatedReward;
+        rewards[rewardIndex] = reward;
+        emit RewardCreated(
+            reward.sponsor,
+            reward.rewardToken,
+            rewardIndex,
+            reward.maximumRewardAmount,
+            reward.earliestProposalTimestamp,
+            reward.optimisticOracleProposerBond,
+            reward.optimisticOracleLivenessTime,
+            reward.priceIdentifier,
+            reward.customAncillaryData
+        );
+
+        // Bump nextCreatedReward index.
+        nextCreatedReward++;
+    }
+
+    /**
+     * @notice Allows anyone to deposit additional rewards for distribution before `earliestProposalTimestamp`.
+     * @dev The caller must approve this contract to transfer `additionalRewardAmount` amount of `rewardToken`.
+     * @param rewardIndex Index for identifying existing rewards object that should receive additional funding.
+     * @param additionalRewardAmount Additional reward amount that the sponsor is posting for distribution.
+     */
+    function increaseReward(uint256 rewardIndex, uint256 additionalRewardAmount) external nonReentrant() {
+        require(rewards[rewardIndex].sponsor != address(0), "Invalid rewardIndex");
+        require(getCurrentTime() < rewards[rewardIndex].earliestProposalTimestamp, "Funding period ended");
+
+        // Pull additional rewards from the sponsor.
+        rewards[rewardIndex].rewardToken.safeTransferFrom(msg.sender, address(this), additionalRewardAmount);
+
+        // Update maximumRewardAmount and log new amount.
+        rewards[rewardIndex].maximumRewardAmount += additionalRewardAmount;
+        emit RewardIncreased(rewardIndex, rewards[rewardIndex].maximumRewardAmount);
+    }
+
+    /********************************************
+     *          DISTRIBUTION FUNCTIONS          *
+     ********************************************/
+
+    /**
+     * @notice Allows any caller to propose distribution for funded reward starting from `earliestProposalTimestamp`.
+     * @dev The caller must approve this contract to transfer `optimisticOracleProposerBond` + final fee amount
+     * of `bondToken`.
+     * @param rewardIndex Index for identifying existing rewards object that should be proposed for distribution.
+     * @param merkleRoot Merkle root describing allocation of proposed rewards distribution.
+     * @param ipfsHash Hash of IPFS object, conveniently stored for clients to verify proposed distribution.
+     */
+    function proposeDistribution(
```

Comment:
a question about the structure of this proposal function:

I am correct in saying that there is always, at maximum, one proposal per `rewardIndex`, right? i.e the proposal is showing how to distribute the reward defined in the `rewardIndex`?

If this is the case, would you the case do you think it makes sence to only allow for ONE proposal to happen at a time? consider the case where two people propose valid `merkleRoot`s right after each other and both are in liveness for the same period. As soon as the first proposal passes liveness without dispute the second proposal will become invalid as you cant double propose for the same thing. This means that, if timed correctly you'd be able to get the second proposal in and passed liveness before it can be disputed.

We have a similar issue in Across-v2 with bundle proposals and the possibility that two bundles will have overlapping information. In that context we fixed by saying that you can only ever have one proposal in liveness at a time. period. you could do the same thing here but restrict it such that you only ever have one proposal _perRewardIndex_ at a time.

The implementation for this would look like adding an enum `DistributionProposed { None, Pending, Accepted }; `within the state of  to `Reward`. This  indicates if there is a current state of the proposal for the given `rewards[rewardIndex]` which then blocks subsequent proposals until either: 1) Set to `Pending` when this method is called thereby blocking other requests (must be on `None` state to accept a proposal) 2) the proposal passes liveness and is accepted via the `executeDistribution` which would set this to `Accepted` (also blocking future proposals) or 3) is called back from the OO via the `priceDisputed` callback function that would re-set this to `None` which will enable future proposals to happen again.

I'm curious what @UMAprotocol/smart-contracts thinks about this proposed change ^^

Reply:
Yeah I think it simplifies the contract a bit to only allow one proposal at a time. The pros are:
- gas savings: only need to store one proposal struct instead of an array
- practicality: I think these rewards will be paid out relatively infrequently, maybe once a day max, so no need to allow multiple to be in liveness at the same time

Reply:
In order to use `priceDisputed` callback from OO to flag `Reward` as `DistributionProposed.None` the callback function would need to parse ancillary data to identify appended `proposalIndex`. Parsing strings in solidity does not sound great, so as alternative approach I now plan to:
- `proposalIndex` replaced with `rewardIndex` when appending to ancillary data
- hash `identifier`, `timestamp` and `ancillaryData`
- use above hash to address `Proposal` structs in a mapping
- in `priceDisputed` callback function hash the passed `identifier`, `timestamp` and `ancillaryData` arguments from OO that allows identifying `Proposal` struct and link it back to relevant `Reward` struct so that it can be flagged as `DistributionProposed.None` and new proposals allowed for the disputed reward.

Reply:
I switched to dynamic array for `Reward` structs and mapped `Proposal` structs by hashing `identifier`, `timestamp` and `ancillaryData` that is looked up in `priceDisputed` callback function.

Now only one undisputed distribution proposal is allowed through enum `DistributionProposed` and I am not deleting `Proposal` structs anymore.

Reply:
excellent! the way you have it now is great@!